### ISSUE TITLE
training speed

### ISSUE THREAD

**Issue by Iamal1** [ruotianluo/ImageCaptioning.pytorch#81]
I find it takes very long to train a model, for example, for show, attend, tell, I set batch_size=64 and it takes  about 16s to read data, and for adaptive attention model, I set batch_size=32 and it takes 8s to read data. So I find that it takes about 0.25s to read the feature map for one image, and it takes about 24 hours for 3 epochs, which means for 25 epochs I need 1 week, is that too long for this?  I'm not familiar with .h5 files and wondering why it takes so long, is it the same for you guys?

**Comment by ruotianluo** [comment]
Usually it will be much faster after first epoch. But still the reading time seems too large.

**Comment by Iamal1** [comment]
Thanks for your answer, I just recognize that my attention size is default 14 by 14, so the cocotalk_att file is almost  400G, much larger than the 7x7 one(but not 4 times?). For st model without attention it is quite fast to finish the 25 epochs, so I think this could make some sense and I might do some trials on a smaller attention.

**Comment by Iamal1** [comment]
@ruotianluo I think I almost figure out this problem. I generate the cocotalk_att with attention size 7x7, and the file size is about 100G, and the training time(data reading time) is almost 1/4 of the 14x14 attention, so the main reason for my long training time is the large attention size.  Besides, the machine also has impact on the results. I run the same code on a newer machine and get surprising results. Using 7x7 attention for sat only takes 1 hour to train 5 epochs(more than 20x faster than the old machine), while 14x14 will probably take 2-3 days for 25 epochs, still kind of slow(several seconds for batch_size 64). As a result, I will carry out my experiment on the better machine and thank you again for your contribution.

**Comment by hhxxttxsh** [comment]
@Iamal1 When you use 7x7 attention, does it affect the final accuracy?
@ruotianluo I am currently using the same setting like the one in the author's latest commit for training the topdown model. The batch_size = 1 and the num_workers=multiprocessing.cpu_count() (in my machine, it is 56) in the dataloader constructor:
        self.split_loader = iter(
            data.DataLoader(dataset=self.dataloader,
                            batch_size=1,
                            sampler=sampler,
                            shuffle=False,
                            pin_memory=True,
                            num_workers=multiprocessing.cpu_count(),
                            collate_fn=lambda x: x[0]))
For each iteration, the reading data process during training takes 4-12 s, and that is really slow. I tried to change the batch_size here and the num_workers, but it only made it even slower. Any suggestions on this?
